pkg/process/events/listener_linux.go

```diff
@@ -0,0 +1,165 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// Copyright 2016-present Datadog, Inc.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+
+//go:build linux
+// +build linux
+
+package events
+
+import (
+	"context"
+	"errors"
+	"go.uber.org/atomic"
+	"io"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/cenkalti/backoff"
+	"google.golang.org/grpc"
+
+	ddconfig "github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/process/events/model"
+	"github.com/DataDog/datadog-agent/pkg/security/api"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// SysProbeListener collects process events using the runtime-security module in system-probe
+type SysProbeListener struct {
+	client        api.SecurityModuleClient
+	conn          *grpc.ClientConn
+	running       atomic.Value
+	connected     atomic.Value
+	retryInterval time.Duration
+	handler       EventHandler
+	wg            sync.WaitGroup
+}
+
+// NewListener returns a new SysProbeListener to listen for process events
+func NewListener(handler EventHandler) (*SysProbeListener, error) {
+	socketPath := ddconfig.Datadog.GetString("runtime_security_config.socket")
+	if socketPath == "" {
+		return nil, errors.New("runtime_security_config.socket must be set")
+	}
+
+	conn, err := grpc.Dial(socketPath, grpc.WithInsecure(), grpc.WithContextDialer(func(ctx context.Context, url string) (net.Conn, error) {
+		return net.Dial("unix", url)
+	}))
+	if err != nil {
+		return nil, err
+	}
+
+	client := api.NewSecurityModuleClient(conn)
+	return newSysProbeListener(conn, client, handler)
+}
+
+// newSysProbeListener returns a new SysPobeListener
+func newSysProbeListener(conn *grpc.ClientConn, client api.SecurityModuleClient, handler EventHandler) (*SysProbeListener, error) {
+	if handler == nil {
+		return nil, errors.New("can't create a Listener without an EventHandler")
+	}
+
+	return &SysProbeListener{
+		client:        client,
+		conn:          conn,
+		retryInterval: 2 * time.Second,
+		handler:       handler,
+	}, nil
+}
+
+// Run starts a new thread to listen for process events
+func (l *SysProbeListener) Run() {
+	l.wg.Add(1)
+	go func() {
+		defer l.wg.Done()
+		l.run()
+	}()
+}
+
+// run keeps polling the SecurityModule server for process events
+func (l *SysProbeListener) run() {
+	log.Info("Start listening for process events")
+	l.running.Store(true)
```

Comment:
It looks like there may be a rare possibility where `Stop()` is called right after run() and may set `running` to `false` before this line is reached and the listener won't stop. Maybe we should have this `l.running.Store(true)` happen in the `Run()` method on the main thread instead.

Reply:
Oh, you're completely right! Thanks for catching it.
Fixed in https://github.com/DataDog/datadog-agent/pull/12099/commits/3483bcc299ebda145257a11d10995abdb11148af .

Reply:
In fact, I've ended up dropping the atomic `running` and using go channels to control the main routine loop: https://github.com/DataDog/datadog-agent/pull/12099/commits/85ba8f25597a3970402750aad60211ed9558c3e3.